File: modules/rust_workloads/mem/mem_retained_growth.rs
//! One-shot retained-growth workload: the heap profile attributes the whole
//! live heap to the growth site.

mod prof_dump;

use rust_workloads::{retained_growth, workload_n};

const DEFAULT_CHUNKS: usize = 65_536;
const CHUNK_BYTES: usize = 1024;

fn main() {
    let chunks = workload_n(DEFAULT_CHUNKS);
    let retained = retained_growth::grow(chunks, CHUNK_BYTES);
    let out = prof_dump::dump();
    println!(
        "retained {} bytes in {} chunks; heap profile: {}",
        retained_growth::retained_bytes(&retained),
        retained.len(),
        out.display()
    );
}

File: modules/cpp_workloads/mem/mem_retained_growth.cpp
// One-shot retained-growth workload: the heap profile attributes the
// whole live heap to the growth site.

#include <cstddef>
#include <cstdio>

#include "prof_dump.h"
#include "retained_growth.h"
#include "workload_n.h"

namespace {
constexpr std::size_t DEFAULT_CHUNKS = 65536;
constexpr std::size_t CHUNK_BYTES = 1024;
}  // namespace

auto main() -> int {
    const std::size_t chunks = cpp_workloads::workload_n(DEFAULT_CHUNKS);
    cpp_workloads::heap_profile_start();
    const auto retained = cpp_workloads::grow(chunks, CHUNK_BYTES);
    const auto out = cpp_workloads::heap_profile_dump();
    std::printf("retained %zu bytes in %zu chunks; heap profile prefix: %s\n",
                cpp_workloads::retained_bytes(retained), retained.size(), out.c_str());
    return 0;
}
